### ISSUE TITLE
[Performance] Cholesky operation

### ISSUE THREAD

**Issue by awav** [tensorflow/swift-apis#568]
Hello everyone,

I have a concern about this line: https://github.com/tensorflow/swift-apis/blob/master/Sources/TensorFlow/Operators/Math.swift#L2712. In TensorFlow for python - gradients have access to the input and output of the operation, therefore there is no need in recomputing A=LLᵀ. The gradient signature normally looks like `gradient(op, grad)`, where op has attributes like input and output.

But [here](https://github.com/tensorflow/swift-apis/blob/master/Sources/TensorFlow/Operators/Math.swift#L2712), vector jacobian has to recompute decomposition once more time. Can it be avoided somehow?

PS: Actually, there are numerous examples where forward step value is necessary for gradient computation and this is quite important for computationally efficient gradient computation in general.

**Comment by awav** [comment]
Okay, looks like it is my misunderstanding of the `@differentiable`. The tuple returns a function for the gradient and first element is simply a forward step.

Feel free to close the issue, it that's the case :)

**Comment by dan-zheng** [comment]
> Okay, looks like it is my misunderstanding of the `@differentiable`. The tuple returns a function for the gradient and first element is simply a forward step.

This is accurate!

```swift
// Given a function type `(T0, ...) -> U`
// (where `T0`, ..., `U` conform to the `Differentiable` protocol †),
//
// The VJP function typing rules are:
//
//  (T0, ...)      ->  (U,    (U.TangentVector)   -> (T0.TangentVector, ...))
//   ^                  ^      ^~~~~~~~~~~~~~~        ^~~~~~~~~~~~~~~~~~~~~
//  original args   result    derivative wrt result     derivative wrt args
//
// The derivative function returns a tuple of:
// - The original result of type `U`.
// - A "backpropagator" pullback function takes that the derivative with respect to
//   the result and returns derivatives with respect to arguments.
//
// †: only "wrt" parameters need to conform to `Differentiable`.
```

Derivative functions are defined as VJP functions taking original arguments rather than just the "returned pullback" functions. This allows the pullback function to capture and use intermediate values computed in the original function.

That's exactly what happens for `cholesky`: `_vjpCholesky` computes `let decomposition = cholesky(x)`, and the pullback closure captures and uses `decomposition` rather than recomputing it.

Please read the ["JVP and VJP functions" section](https://docs.google.com/document/d/1_BirmTqdotglwNTOcYAW-ib6mx_jl-gH9Dbg4WmHZh0/edit#bookmark=id.262kfivwzyzr) from [bit.ly/swift-autodiff-internals](http://bit.ly/swift-autodiff-internals) for more details!

**Comment by awav** [comment]
gosh, I love that...

**Comment by dan-zheng** [comment]
If you're interested in more cool differentiation stuff, check out our [custom differentiation tutorial](https://github.com/tensorflow/swift/blob/master/docs/site/tutorials/custom_differentiation.ipynb):
* Stop gradients from propagating
* Derivative surgery
* Gradient checkpointing in 6 lines

Other cool features are on the roadmap:
- [Retroactive derivative registration](https://gist.github.com/dan-zheng/0e28c41dd0734fa6d3644e6ed0a9c16f) (in-progress)
  - Register derivatives for functions in other modules, like functions imported from C.
- [Default derivative implementations for protocol requirements](https://bugs.swift.org/browse/TF-982)
  - e.g. so you'll never have to manually define a derivative for [`AdditiveArithmetic.+`](https://developer.apple.com/documentation/swift/additivearithmetic/3126821) or [`ElementaryFunctions.exp`](https://github.com/apple/swift-numerics/tree/master/Sources/Real#protocols-and-methods) again for conforming types of those protocols
